### ISSUE TITLE
'Specified cast is not valid' on IHttpClientFactory.CreateClient

### ISSUE THREAD

**Issue by SagaHuS** [dotnet/aspnetcore#29487]
### Describe the bug
IHttpClientFactory.CreateClient("name") throws an Exception with 'Specified cast is not valid'. But custom code contains no cast

### To Reproduce
1. Create New Blazor Webassembly App in VS 2019 (v16.8.4)
2. Choose .Net 5.0, Authentication Single Account, no https, not core hosted, not progressive
3. Add Microsoft.Extensions.Http Nuget Package (v5.0.0)
4. Add CustomAuthenticationStateProvider
```
public class CustomAuthenticationStateProvider : AuthenticationStateProvider
{
    public override Task<AuthenticationState> GetAuthenticationStateAsync()
        {
            //Login failed
            return Task.FromResult(new AuthenticationState(new ClaimsPrincipal(new ClaimsIdentity())));
        }
}
```
5. Add CustomAuthorizationMessageHandler 
```
public class CustomAuthorizationMessageHandler : AuthorizationMessageHandler
{
    private readonly CustomAuthenticationStateProvider _customAuthenticationStateProvider;

    public CustomAuthorizationMessageHandler(IAccessTokenProvider provider, NavigationManager navigation, CustomAuthenticationStateProvider customAuthenticationStateProvider) : base(provider, navigation)
    {
        if (customAuthenticationStateProvider == null)
        {
            throw new ArgumentNullException(nameof(customAuthenticationStateProvider));
        }
        _customAuthenticationStateProvider = customAuthenticationStateProvider;
    }
}
```
6. Modify LoginDisplay.razor (inject HttpClientFactory, button unter NotAuthorized, Test-Method)
```
@using Microsoft.AspNetCore.Components.Authorization
@using Microsoft.AspNetCore.Components.WebAssembly.Authentication

@inject NavigationManager Navigation
@inject SignOutSessionStateManager SignOutManager
@inject IHttpClientFactory HttpClientFactory

<AuthorizeView>
    <Authorized>
        Hello, @context.User.Identity.Name!
        <button class="nav-link btn btn-link" @onclick="BeginSignOut">Log out</button>
    </Authorized>
    <NotAuthorized>
        <button @onclick="Test">Test</button>
    </NotAuthorized>
</AuthorizeView>

@code{
    private async Task BeginSignOut(MouseEventArgs args)
    {
        await SignOutManager.SetSignOutState();
        Navigation.NavigateTo("authentication/logout");
    }

    private async Task Test(MouseEventArgs e)
    {
        using(HttpClient client = HttpClientFactory.CreateClient("ServerAPI"))
        {
            ...
        }
    }
}
```
7. Modify Program.Startup
```
    var builder = WebAssemblyHostBuilder.CreateDefault(args);
    builder.RootComponents.Add<App>("#app");            

    builder.Services.AddApiAuthorization();
    builder.Services.AddAuthorizationCore();

    builder.Services.AddScoped<CustomAuthenticationStateProvider>();
    builder.Services.AddScoped<AuthenticationStateProvider, CustomAuthenticationStateProvider>();
    builder.Services.AddScoped<CustomAuthorizationMessageHandler>();
    builder.Services.AddScoped<AuthorizationMessageHandler, CustomAuthorizationMessageHandler>();

    builder.Services.AddHttpClient("ServerAPI").AddHttpMessageHandler(sp =>
    {
        return sp.GetService<CustomAuthorizationMessageHandler>().ConfigureHandler(authorizedUrls: new[] { "http://localhost:10000" });
    });

    await builder.Build().RunAsync();
```
8. Click Test Button and get error

### Exceptions (if any)
```
System.InvalidCastException: Specified cast is not valid.
   at Microsoft.Extensions.DependencyInjection.WebAssemblyAuthenticationServiceCollectionExtensions.<>c__0`3[[Microsoft.AspNetCore.Components.WebAssembly.Authentication.RemoteAuthenticationState, Microsoft.AspNetCore.Components.WebAssembly.Authentication, Version=5.0.2.0, Culture=neutral, PublicKeyToken=adb9793829ddae60],[Microsoft.AspNetCore.Components.WebAssembly.Authentication.RemoteUserAccount, Microsoft.AspNetCore.Components.WebAssembly.Authentication, Version=5.0.2.0, Culture=neutral, PublicKeyToken=adb9793829ddae60],[Microsoft.AspNetCore.Components.WebAssembly.Authentication.ApiAuthorizationProviderOptions, Microsoft.AspNetCore.Components.WebAssembly.Authentication, Version=5.0.2.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]].<AddRemoteAuthentication>b__0_1(IServiceProvider sp)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2[[Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext, Microsoft.Extensions.DependencyInjection, Version=5.0.0.1, Culture=neutral, PublicKeyToken=adb9793829ddae60],[System.Object, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].VisitCallSiteMain(ServiceCallSite callSite, RuntimeResolverContext argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2[[Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext, Microsoft.Extensions.DependencyInjection, Version=5.0.0.1, Culture=neutral, PublicKeyToken=adb9793829ddae60],[System.Object, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].VisitCallSite(ServiceCallSite callSite, RuntimeResolverContext argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2[[Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext, Microsoft.Extensions.DependencyInjection, Version=5.0.0.1, Culture=neutral, PublicKeyToken=adb9793829ddae60],[System.Object, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].VisitCallSiteMain(ServiceCallSite callSite, RuntimeResolverContext argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2[[Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext, Microsoft.Extensions.DependencyInjection, Version=5.0.0.1, Culture=neutral, PublicKeyToken=adb9793829ddae60],[System.Object, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].VisitCallSite(ServiceCallSite callSite, RuntimeResolverContext argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__1(ServiceProviderEngineScope p)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[CustomAuthorizationMessageHandler](IServiceProvider provider)
   at DataLayerGeneratorClient.Program.<>c.<Main>b__0_0(IServiceProvider sp) in D:\Dev\4.0\Test2\DataLayerGenerat…tem.Lazy`1[[Microsoft.Extensions.Http.ActiveHandlerTrackingEntry, Microsoft.Extensions.Http, Version=5.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]].ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1[[Microsoft.Extensions.Http.ActiveHandlerTrackingEntry, Microsoft.Extensions.Http, Version=5.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]].CreateValue()
   at System.Lazy`1[[Microsoft.Extensions.Http.ActiveHandlerTrackingEntry, Microsoft.Extensions.Http, Version=5.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]].get_Value()
   at Microsoft.Extensions.Http.DefaultHttpClientFactory.CreateHandler(String name)
   at Microsoft.Extensions.Http.DefaultHttpClientFactory.CreateClient(String name)
   at DataLayerGeneratorClient.Program.<>c.<Main>b__0_1(IServiceProvider serviceProvider) in D:\Dev\4.0\Test2\DataLayerGeneratorClient\Program.cs:line 32
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2[[Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext, Microsoft.Extensions.DependencyInjection, Version=5.0.0.1, Culture=neutral, PublicKeyToken=adb9793829ddae60],[System.Object, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].VisitCallSiteMain(ServiceCallSite callSite, RuntimeResolverContext argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2[[Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext, Microsoft.Extensions.DependencyInjection, Version=5.0.0.1, Culture=neutral, PublicKeyToken=adb9793829ddae60],[System.Object, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].VisitCallSite(ServiceCallSite callSite, RuntimeResolverContext argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__1(ServiceProviderEngineScope p)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.AspNetCore.Components.ComponentFactory.<>c__DisplayClass6_0.<CreateInitializer>g__Initialize|2(IServiceProvider serviceProvider, IComponent component)
   at Microsoft.AspNetCore.Components.ComponentFactory.PerformPropertyInjection(IServiceProvider serviceProvider, IComponent instance)
   at Microsoft.AspNetCore.Components.ComponentFactory.InstantiateComponent(IServiceProvider serviceProvider, Type componentType)
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.InstantiateComponent(Type componentType)
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.InstantiateChildComponentOnFrame(RenderTreeFrame& frame, Int32 parentComponentId)
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.InitializeNewComponentFrame(DiffContext& diffContext, Int32 frameIndex)
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.InitializeNewSubtree(DiffContext& diffContext, Int32 frameIndex)
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.InsertNewFrame(DiffContext& diffContext, Int32 newFrameIndex)
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForRange(DiffContext& diffContext, Int32 oldStartIndex, Int32 oldEndIndexExcl, Int32 newStartIndex, Int32 newEndIndexExcl)
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.ComputeDiff(Renderer renderer, RenderBatchBuilder batchBuilder, Int32 componentId, ArrayRange`1 oldTree, ArrayRange`1 newTree)
   at Microsoft.AspNetCore.Components.Rendering.ComponentState.RenderIntoBatch(RenderBatchBuilder batchBuilder, RenderFragment renderFragment)
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.RenderInExistingBatch(RenderQueueEntry renderQueueEntry)
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.ProcessRenderQueue()…
```

### Further technical details
- Include the output of `dotnet --info`
```.NET SDK (gemäß "global.json"):
 Version:   5.0.102
 Commit:    71365b4d42

Laufzeitumgebung:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.102\

Host (useful for support):
  Version: 5.0.2
  Commit:  cb5f173b96

.NET SDKs installed:
  5.0.102 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download
```
- Visual Studio 2019 v16.8.4

**Comment by javiercn** [comment]
@SagaHuS thanks for contacting us.

You can't customize `AuthenticationStateProvider` without also updating the registrations for other  `Microsoft.AspNetCore.Components.WebAssembly.Authentication` services.

In general, you shouldn't need to implement your own `AuthenticationStateProvider` if you are using `Microsoft.AspNetCore.Components.WebAssembly.Authentication`.

If you tell us more about what you are trying to achieve we might be able to suggest alternatives.

**Comment by SagaHuS** [comment]
@javiecrn thanks for the quick answer
How can one update registrations for other Microsoft.AspNetCore.Components.WebAssembly.Authentication services?

I am trying to implement so very basic authentication and authorization scheme without server behind. And I want AuthorizeView and AuthorizeRouteView to work with it. So I thought it would be the easiest way to implement a custom AuthenticationStateProvider. I have found lots of examples online where other people were doing it, too

**Comment by javiercn** [comment]
@SagaHuS `Microsoft.AspNetCore.Components.WebAssembly.Authentication` is heavily based towards token based protocols. 

If you are not doing token based auth, then I would suggest you avoid this package completely if you really want to go that route, since you don't need anything on it, and avoid the AuthorizationMessageHandler all together since its designed to be used with it for attaching tokens (which I don't think you are doing).

All that said, I would **strongly** recommend against implementing your own authentication. It might seem simple, but getting it right is hard and the consequences are bad.

**Comment by SagaHuS** [comment]
@javiercn Thanks again. Right now there are no tokens, but that will most likely change.
I got my ideas mostly from https://docs.microsoft.com/en-us/aspnet/core/blazor/security/?view=aspnetcore-5.0
There is also suggested to implement a custom AuthenticationStateProvider (like on soo many other sites and posts). That seems pretty straight forward and I really would like to use it. The implementation I posted worked. But then I also wanted to start using IHttpClientFactory, which does NOT work when there is custom AuthenticationStateProvider. Are you sure this is NOT a bug? Should they really exclude each other?

**Comment by javiercn** [comment]
@SagaHuS I'm sure (I wrote it).

And the reason is because you are calling `AddApiAuthorization` but then are choosing to override only a few of the provided services. If you want to use something different you shouldn't be calling AddApiAuthorization at all and if you do but want to override the authenticationstateprovider you need to also provide alternative implementations for all the services that rely on it.

`IAccessTokenProvider` for example. The issue is that you are using the package when you don't need it (for your purposes) and you are not providing an implementation that works in concert with the one you override for some other dependent services.

I'm very interested in understanding why are you trying to implement your own authentication, specially if you are using tokens all the functionality we provide should already work out of the box.

Is there any reason why our stock implementation doesn't work for your case?

**Comment by SagaHuS** [comment]
@javiercn 
Right now in our company there is the decision to make if we switch from WPF to Blazor WASM. I am trying to examine all possibilities with this new framework. I would like to get a deep understanding of how everything works.
You mentioned IAccessTokenProvider. What other services would need a custom implementation?
No the stock implementation may suite us, but so far I do not understand everthing and cant judge if it is the best way in the end

**Comment by SagaHuS** [comment]
@javiernc okay I understand.
By now I got it working. I do not really understand what AddApiAuthorization does in detail. But since I implemented my own CustomAuthenticationStateProvider in my understanding I dont need it anymore. What I have also changed is that instead of a CustomAuthorizationMessageHandler I implemented a custom CustomHttpMessageHandler inherting from DelegatingHandler. That is where I attach my token recieved from login.

I guess my initial problem could named as "Unprecise error message for an unintended situation". But who am I to judge.
Anyway many many thanks for your help! I am closing the issue
